onmt/translate/Translator.lua

```diff
@@ -125,179 +123,100 @@ function Translator:translateBatch(batch)
     goldScore = self.models.decoder:computeScore(batch, encStates, context)
   end
 
-  -- Expand tensors for each beam.
-  context = context
-    :contiguous()
-    :view(1, batch.size, batch.sourceLength, self.checkpoint.options.rnn_size)
-    :expand(self.opt.beam_size, batch.size, batch.sourceLength, self.checkpoint.options.rnn_size)
-    :contiguous()
-    :view(self.opt.beam_size * batch.size, batch.sourceLength, self.checkpoint.options.rnn_size)
-
-  for j = 1, #encStates do
-    encStates[j] = encStates[j]
-      :view(1, batch.size, self.checkpoint.options.rnn_size)
-      :expand(self.opt.beam_size, batch.size, self.checkpoint.options.rnn_size)
-      :contiguous()
-      :view(self.opt.beam_size * batch.size, self.checkpoint.options.rnn_size)
-  end
-
-  local remainingSents = batch.size
-
-  -- As finished sentences are removed from the batch, this table maps the batches
-  -- to their index within the remaining sentences.
-  local batchIdx = {}
-
-  local beam = {}
-
-  for b = 1, batch.size do
-    table.insert(beam, onmt.translate.Beam.new(self.opt.beam_size, #self.dicts.tgt.features))
-    table.insert(batchIdx, b)
-  end
-
-  local i = 1
-
-  local decOut
-  local decStates = encStates
-
-  while remainingSents > 0 and i < self.opt.max_sent_length do
-    i = i + 1
-
-    -- Prepare decoder input.
-    local input = torch.IntTensor(self.opt.beam_size, remainingSents)
-    local inputFeatures = {}
-    local sourceSizes = torch.IntTensor(remainingSents)
-
-    for b = 1, batch.size do
-      if not beam[b].done then
-        local idx = batchIdx[b]
-        sourceSizes[idx] = batch.sourceSize[b]
-
-        -- Get current state of the beam search.
-        local wordState, featuresState = beam[b]:getCurrentState()
-        input[{{}, idx}]:copy(wordState)
-
-        for j = 1, #self.dicts.tgt.features do
-          if inputFeatures[j] == nil then
-            inputFeatures[j] = torch.IntTensor(self.opt.beam_size, remainingSents)
-          end
-          inputFeatures[j][{{}, idx}]:copy(featuresState[j])
-        end
+  -- prepare inputs for next time step
+  local function feedFunction(stepOutputs, topIndexes)
+    if stepOutputs == nil then -- initial inputs for first time step
+      local input = onmt.utils.Cuda.convert(torch.IntTensor(batch.size)):fill(onmt.Constants.BOS)
```

Comment:
I'm using a new `localize` function instead, which makes sure the consistency between two tensors. (I have to convert back and forth due to some functions in DoubleTensor are not implemented in ByteTensor or LongTensor.)